Order Fulfillment: Add Optional Shipment Tracking Card

anitaa1990 opened the pull request:
PR 1 of 4. Fixes #1035 by providing an option to add tracking information in order fulfilment screen if the Shipment Tracking plugin is available/installed.

Items completed in the PR:
- [x] `Add Tracking` button added to Tracking card and displayed only in order fulfilment page 
- [x] Added logic to display shipment tracking card in fulfilment page only if tracking info is available. If tracking item is displayed in fulfilment page, then `Track package` button should not be displayed.
- [x] Clicking on the Add Tracking button should redirect to Add Shipment activity
- [x] Add Shipment provider UI
- [x] Tracking number field should just accept text entry and cap the length at 255.
- [x] Date shipped should open a calendar for selection. Default date would be set to current date.
- [x] Clicking on Carrier should open dialog to display carrier list sorted in the following format:
    - [x] the list is sorted alphabetically by country.
- [x] Clicking ADD should add the tracking record to the Optional Tracking card
- [x] Clicking back from Order fulfilment page to Order Detail page after adding a tracking record, the record should be included in the Order Detail page Tracking Card
- [x] Add search functionality to the provider list
- [x] Unit tests for Presenter class
- [x] Add relevant events to the shipment tracking feature
- [x] Bug fix: first section header is not displayed for some reason
- [x] Bug fix: he tracking card is not displayed because the shipment tracking list returns empty when trying again in Order Fulfilment screen
- [x] Snackbar error displaying behind the dialog (when trying to show error message)
- [x] Seems kind of inefficient to fetch data from api each time the dialog is loaded.

### Testing
Unit tests and are included in `AddOrderTrackingProviderListPresenterTest` and `OrderFulfillmentPresenterTest`.

### Screenshots
<img src="https://user-images.githubusercontent.com/22608780/57831228-2a6af180-77d2-11e9-8d45-2d1e8bf64b82.gif" width="300"/>

<img src="https://user-images.githubusercontent.com/22608780/57831218-1d4e0280-77d2-11e9-829c-c5476b3b84b5.gif" width="300"/>

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

cc @nbradbury @AmandaRiu

anitaa1990 commented:
> Snackbar error displaying behind the dialog (when trying to show error message)

Note: need to rethink how to display Snackbar error over the dialog popup. It doesn't seem possible without having to create our [own custom SnackBar](https://stackoverflow.com/questions/45760678/how-does-google-inbox-show-a-snackbar-covering-the-keyboard).
And technically not sure if displaying a SnackBar inside dialog is the right way to go.

> The provider list data is being fetched every time there is an active internet connection. It feels inefficient to fetch data from api each time the dialog is loaded. 

The list of providers is getting loaded every time we open the dialog to select a provider. ([here](https://github.com/woocommerce/woocommerce-android/blob/55f53a42678afa88ceaf09ac88f168a70eb91bb1/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/AddOrderShipmentTrackingPresenter.kt#L70)). This seems unnecessary given that the provider list is getting cached in the local db. Perhaps we can have a mechanism to load providers based on some time limit? Or once every time the app is opened (only if the plugin is available)? Or initiate the request only once when add tracking button is clicked, before the dialog is even opened?

anitaa1990 commented:
@AmandaRiu @nbradbury PR is ready for first round of review 👍

There are a couple of points that I wanted to bring to attention during the review and I thought I would jot them down here.

> Should the selected provider name be remembered across all orders?

It makes sense that the user should like to use the same provider they had selected for a previous order. So it would be better to store this information in shared preferences and then fetch this data and display. At the moment though, storing this info only in memory. So if the user selects a different order, then the provider name they had previously selected would not be remembered and displayed. Perhaps this is something that can be enhanced when tackling this issue #1058 ?

> Shipment tracking provider list is fetched once per order at the moment

So when user selects another order, and clicks on the Add Tracking button, the provider list will be fetched again from api. Again not sure if this makes sense given that the provider list is not order specific.

AmandaRiu commented:
@nbradbury @anitaa1990 

> So when user selects another order, and clicks on the Add Tracking button, the provider list will be fetched again from api. Again not sure if this makes sense given that the provider list is not order specific.

My plan was to only fetch the providers once during a single session.

nbradbury commented:
> My plan was to only fetch the providers once during a single session.

I agree that sounds better.

nbradbury commented:
> Perhaps this is something that can be enhanced when tackling this issue #1058 ?

I think that's fine, but shouldn't this PR be merged into a feature branch rather than `develop` since the feature won't completed until #1058 is done?

nbradbury commented:
I'm noticing some minor differences from the blueprint. In the blueprint (left), there are no arrows next to the provider or date, but there are in this PR (right).

![Untitled](https://user-images.githubusercontent.com/3903757/57635797-e380c000-7575-11e9-8843-c40638188617.png)

Also, note that the first time this screen appears the provider is blank, which looks odd to me. Perhaps some grayed placeholder text, like "Select provider"? <- @kyleaparker

nbradbury commented:
There doesn't appear to be a way to edit or remove a tracking. I expected to be able to tap a tracking item to change it, but that copies the tracking number to the clipboard instead. Is this coming in a subsequent PR?

![Screenshot_1557763802](https://user-images.githubusercontent.com/3903757/57636892-1c219900-7578-11e9-809d-986d67dcc78f.png)

nbradbury commented:
To me it's a bit odd that to select a provider, you have to tap it and then hit the back button. Shouldn't simply tapping it be sufficient?

![Screenshot_1557763891](https://user-images.githubusercontent.com/3903757/57637020-5c811700-7578-11e9-825b-2a91edc293d2.png)

nbradbury commented:
Looks like search isn't very usable in landscape. I think we can fix this by adding:
```
searchView?.let {
     it.imeOptions = it.imeOptions or EditorInfo.IME_FLAG_NO_EXTRACT_UI
}
```

![Screenshot_1557764025](https://user-images.githubusercontent.com/3903757/57637555-b209f380-7579-11e9-83bb-8d407d4b18de.png)

nbradbury commented on WooCommerce/src/main/AndroidManifest.xml:
```diff
@@ -74,6 +74,10 @@
             android:name=".ui.imageviewer.ImageViewerActivity"
             android:theme="@style/ImageViewer.Activity"/>
 
+        <activity
```
I'm not a big fan of using `configChanges` this way. We do that in the main activity, which required overriding `onConfigurationChanged()` in several places where otherwise we wouldn't have needed to. @AmandaRiu Any thoughts on this?

anitaa1990 replied:
So I had this issue where when the provider list dialog is displayed and screen is rotated, the new value selected in the list is not updated in the provider field of the activity. The method in question is called [here](https://github.com/woocommerce/woocommerce-android/blob/anitaa/1035-optional-shipment-tracking/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/AddOrderShipmentTrackingActivity.kt#L185) but the `Button` text is not updated. 

![ezgif com-video-to-gif (1)](https://user-images.githubusercontent.com/22608780/57671407-45eac680-7631-11e9-9de7-847c4fbe06c4.gif)


I agree this is not the best solution and it doesn’t solve the problem. I will look for alternate solutions to fix this.

nbradbury replied:
> I spent an unreasonable amount of time trying to fix this and in the end added a quick fix.

I can so relate to that :)

AmandaRiu replied:
Clicking on an item and immediately closing this view like @nbradbury suggested will fix this issue 👍 So I wouldn't spend anymore time on it. Definitely remove the `configChanges` code.

anitaa1990 replied:
I thought it might be related to that FluxC issue. That issue was because the listener being passed to the dialog would be `null` after a configuration change so the callback to the activity/parent fragment was not happening. But in this case, the callback is being received by the activity and the method in question is called [here](https://github.com/woocommerce/woocommerce-android/blob/anitaa/1035-optional-shipment-tracking/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/AddOrderShipmentTrackingActivity.kt#L185) but the `Button` text is not updated. 

Modifying the provider list dialog to close immediately after selecting an item did not solve it :disappointed:  
<img src="https://user-images.githubusercontent.com/22608780/57751174-7fd7cd80-7702-11e9-8d20-4501497da6af.gif" width="300"/>

AmandaRiu replied:
@anitaa1990  I'm not sure I understand your gif above. I'm not seeing you clicking on anything.  Can you enable `Show taps` in the developer options of your test device? I tested it out on an emulator running API 28 with and without `Don't keep Activities` enabled and it worked seamlessly no matter the option, although logic tells me it shouldn't work. Instead of passing the activity as the listener when you create the fragment, what you could do instead is set the activity as the listener in `onAttach` and remove it as the listener in `onDetach`. Example:
```
override fun onAttach(context: Context?) {
    super.onAttach(context)

    if (activity is AddOrderTrackingProviderActionListener) {
        listener = activity as AddOrderTrackingProviderActionListener
    }
}
```

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/AddOrderShipmentTrackingActivity.kt:
```diff
@@ -0,0 +1,194 @@
+     * Reusing the same menu used for adding order notes
+     */
+    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
+        menuInflater.inflate(R.menu.menu_add_note, menu)
```
Minor nit: how about we rename this menu to `menu_add` since it's now being used outside of order notes?

anitaa1990 replied:
👍

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/AddOrderShipmentTrackingPresenter.kt:
```diff
@@ -0,0 +1,125 @@
+                    addTrackingProviderView?.showSkeleton(true)
+                    requestShipmentTrackingProvidersFromApi(order)
+                } else {
+                    isTrackingProviderFetched = false
```
We already know `isTrackingProviderFetched` is false so this isn't necessary. More importantly, I think this is missing the edge case when there is no connection and providers haven't ever been fetched.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderFulfillmentFragment.kt:
```diff
@@ -82,6 +96,33 @@ class OrderFulfillmentFragment : Fragment(), OrderFulfillmentContract.View, View
     }
 
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        if (requestCode == REQUEST_CODE_ADD_TRACKING) {
```
Should we also be checking `resultCode == Activity.RESULT_OK` here?

anitaa1990 commented:
> My plan was to only fetch the providers once during a single session.

Ah, that makes more sense. I will modify the logic to fetch data from api only if there is nothing in local cache.

> I think that's fine, but shouldn't this PR be merged into a feature branch rather than develop since the feature won't completed until #1058 is done?

Definitely! I am sorry to say I wasn't really aware we could do this and ended up opening a PR against the `develop` branch. I have created a new feature branch `feature/add-shipment-tracking` and modified the PR to merge with that branch.

> I'm noticing some minor differences from the blueprint. In the blueprint (left), there are no arrows next to the provider or date, but there are in this PR (right).
![Untitled](https://user-images.githubusercontent.com/3903757/57635797-e380c000-7575-11e9-8843-c40638188617.png)

This was my bad. I took reference from how iOS had developed it as well.
<img src="https://user-images.githubusercontent.com/22608780/57668816-4af64880-7626-11e9-9205-3721e8f0e1d8.gif" width="300"/>

 I thought the arrow conveyed that you would need to select a provider rather than type. But adding a hint saying `Select Provider` would do the trick as well. @kyleaparker would love your thoughts on this?

> There doesn't appear to be a way to edit or remove a tracking. I expected to be able to tap a tracking item to change it, but that copies the tracking number to the clipboard instead. Is this coming in a subsequent PR?

Based on the discussion from this [P2 post](https://hogsmeadep2.wordpress.com/2019/04/10/shipment-tracking-ui-design-discussion/), I think we removed the option to edit the tracking since this is not supported by the api. Instead, we are providing an option to delete and add again. The option to delete is added as a separate issue #1037 .

> To me it's a bit odd that to select a provider, you have to tap it and then hit the back button. Shouldn't simply tapping it be sufficient?
![Screenshot_1557763891](https://user-images.githubusercontent.com/3903757/57637020-5c811700-7578-11e9-825b-2a91edc293d2.png)

I thought since iOS was implemented this way, it would be consistent to follow this approach. But I agree, this would mean one less click.  Will modify this.

kyleaparker commented:
>Also, note that the first time this screen appears the provider is blank, which looks odd to me. Perhaps some grayed placeholder text, like "Select provider"? <- @kyleaparker

Yes let's add "Select provider" as a placeholder 👍 

>To me it's a bit odd that to select a provider, you have to tap it and then hit the back button. Shouldn't simply tapping it be sufficient?

Agree, it should automatically go back to the previous screen once selected. This should be the same on iOS cc @ctarda

AmandaRiu commented:
> I apologise if this seems like a novice question but wondering what is considered as a session in this case? Is it till the user logs out or till the app is closed and reopened again?

For this app I consider a session to be anytime the app is opened. Reading over your steps it seems the way you have it currently setup seems fine. Does the `isTrackingProviderFetched` variable keep its value even if the app is in the background but still open? Does it retain its value when navigating between orders? If it does, then that would work fine.

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/AddOrderShipmentTrackingPresenter.kt:
```diff
@@ -0,0 +1,132 @@
+    override fun loadShipmentTrackingProviders() {
+        orderModel?.let { order ->
+            if (isTrackingProviderFetched) {
+                loadShipmentTrackingProvidersFromDb()
```
What if we were to just always load from the db and then fetch in the background if a fetch hasn't already happened for this session?

anitaa1990 replied:
Yup, that's what been added in 6d2932d. A flag called `isShipmentTrackingProviderFetched` in `OrderListPresenter` class. When orders are loaded and displayed, we fetch the provider list from api (if `isShipmentTrackingProviderFetched` is false and if network is connected). Then in `AddOrderShipmentTrackingPresenter` we fetch the provider list from local db and if available, then display. Otherwise if network is connected, fetch from api and display. If no network is connected, then display error snack.

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/AddOrderTrackingProviderListFragment.kt:
```diff
@@ -0,0 +1,185 @@
+        toolbar.setNavigationOnClickListener { dismiss() }
+        toolbar.inflateMenu(R.menu.menu_search)
+        val searchMenuItem = toolbar.menu?.findItem(R.id.menu_search)
+        searchView = searchMenuItem?.actionView as SearchView?
```
I recommend using `as?` instead of `as` since that would make this cast null-safe

anitaa1990 replied:
Done in 1267c81

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderFulfillmentPresenter.kt:
```diff
@@ -26,12 +53,80 @@ class OrderFulfillmentPresenter @Inject constructor(
+     * does it make sense to fetch from api again here, even if network is connected?
+     * Unless the user clicks on `Fulfil Order` button before the api data is returned, in which
+     * case it would be better to call api again.
+     * So logic should be to load from cache is available.
```
I agree we should load from cache here since we would've already fetched this information in the order detail screen.

anitaa1990 replied:
Done in f4d7910

AmandaRiu commented on WooCommerce/src/main/res/layout/activity_add_shipment_tracking.xml:
```diff
@@ -0,0 +1,102 @@
+<RelativeLayout
```
Why not use `ConstraintLayout` for a more optimized flattened layout?

anitaa1990 replied:
Definitely. I forgot I had left this for the very end, thanks for pointing it out!

AmandaRiu commented on WooCommerce/src/main/res/values/strings.xml:
```diff
@@ -228,6 +230,21 @@
+    <string name="order_shipment_tracking_date_label">Date shipped</string>
+    <string name="order_shipment_tracking_provider_toolbar_title">Shipping Providers</string>
+    <string name="order_shipment_tracking_provider_list_item">Selected Shipment Provider</string>
+    <string name="orders_shipment_tracking_provider_list_empty_message_with_search">No matching providers</string>
```
np: some of these strings are named with `order_` and others are named with `orders_`, I think just `order_` makes more sense.

anitaa1990 replied:
Done in 69976bc

kyleaparker commented:
>Should we allow the Shipped date date picker to choose a date in the future? (cc: @kyleaparker )

Yes I think we should allow this, shipping can be scheduled for days in the future. We should also allow past dates to allow for packages that were sent before they were fulfilled in the app.

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/analytics/AnalyticsTracker.kt:
```diff
@@ -134,6 +134,16 @@ class AnalyticsTracker private constructor(private val context: Context) {
+        ORDER_SHIPMENT_TRACKING_ADD,
+        ORDER_SHIPMENT_TRACKING_ADD_SUCCESS,
+        ORDER_SHIPMENT_TRACKING_ADD_FAILED,
+        ORDER_SHIPMENT_TRACKING_PROVIDERS_LIST_SEARCH,
```
My bad! I didn't look at the Track Events sheet before adding the events. I have modified this and added events only from the sheet after confirming with Cesar. Done in 0bf6fd7

AmandaRiu replied:
Thanks for updating! I just wrote a post about the naming of these events so this will likely change again in the next few days. FYI.

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/analytics/AnalyticsTracker.kt:
```diff
@@ -134,6 +134,16 @@ class AnalyticsTracker private constructor(private val context: Context) {
         // -- Order Fulfillment
         SNACK_ORDER_MARKED_COMPLETE_UNDO_BUTTON_TAPPED,
         ORDER_FULFILLMENT_MARK_ORDER_COMPLETE_BUTTON_TAPPED,
+        ORDER_FULFILMENT_ORDER_SHIPMENT_TRACKING_LOADED,
```
Nice catch! Modified the events to replicate the ones mentioned in the Events Tracker sheet. Related commit: 0bf6fd7

anitaa1990 commented:
> Yes let's add "Select provider" as a placeholder 👍 no arrow is needed on Android

Added in bfc061f

> For this app I consider a session to be anytime the app is opened. Reading over your steps it seems the way you have it currently setup seems fine. Does the isTrackingProviderFetched variable keep its value even if the app is in the background but still open? Does it retain its value when navigating between orders? If it does, then that would work fine.

Thanks for the clarification @AmandaRiu 👍 . The `isTrackingProviderFetched` variable is initialised in the `OrderFulfilmentPresenter` so it is order specific, at the moment. I have migrated the same logic to `OrderListPresenter` class so as to retain the fetch provider list logic across orders. Added in commit 6d2932d.

> I noticed in the designs the divider line does not go from edge to edge but instead aligns with padding like the Payments card. Currently the order shipment tracking card shows a divider that is edge to edge. I realize this exists prior to this PR, but with the ADD TRACKING button it becomes really cut up looking so I wanted to mention it so we can have it fixed.

Done in 09e1e79

anitaa1990 commented:
Still to complete in the PR before review:

1. Pressing 'X' to cancel closes the view without saving. Should we pop a message asking the user to confirm they want to close without saving the tracking information like the Add Note view does? The same logic should be applied to the back device button as well.

2. Clicking ADD before filling out all fields just ignores the button click. We should have a check here that highlights the field that is required if the user tries to save the tracking info without all the fields populated. This should happen on the individual view like the login screen does if you don't fill out a required field.

3. Load shipment trackings in Order fulfilment page, from cache here since we would've already fetched this information in the order detail screen.

4. Fix button text not getting updated in activity after orientation change and remove `onConfigurationChanged()` from `AndroidManifest.xml`

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/AddOrderShipmentTrackingPresenter.kt:
```diff
@@ -0,0 +1,126 @@
+                    addTrackingProviderView?.showSkeleton(true)
+                    requestShipmentTrackingProvidersFromApi(order)
+                } else {
+                    addTrackingProviderView?.showProviderListErrorSnack()
```
Following this all the way through the user will just see a generic "Error fetching providers" messages if not connected. Since we know the reason we should display something more meaningful such the `offline_error` in `strings.xml` and leave this other error message for actual errors.

anitaa1990 replied:
👍 Fixed in 7e940a7

AmandaRiu commented:
>  Should we pop a message asking the user to confirm they want to close without saving the tracking information like the Add Note view does? 
yup, that sounds perfect.

anitaa1990 commented:
> Pressing 'X' to cancel closes the view without saving. Should we pop a message asking the user to confirm they want to close without saving the tracking information like the Add Note view does? The same logic should be applied to the back device button as well.

Done in 170b3b4


> Clicking ADD before filling out all fields just ignores the button click. We should have a check here that highlights the field that is required if the user tries to save the tracking info without all the fields populated. This should happen on the individual view like the login screen does if you don't fill out a required field.

Done in 39ed3a4

> Load shipment trackings in Order fulfilment page, from cache here since we would've already fetched this information in the order detail screen.

Added logic to pass a boolean flag `isUsingCachedShipmentTrackings` from the `OrderDetailFragment` to the `OrderFulfillmentFragment`. This flag confirms whether the shipment tracking list has already been fetched from api. If value is false, then it will be loaded from cache. If it true, then only if network connected, the shipment tracking list will be fetched from api. Added in f4d7910

> Fix button text not getting updated in activity after orientation change and remove onConfigurationChanged() from AndroidManifest.xml

Fixed in aa591ed

@AmandaRiu @nbradbury PR is ready for another round of review.
@kyleaparker Forgot to add the gif for the PR before 🤦‍♀ Added this now! PR is ready for design review.

nbradbury commented:
The carrier should have `background="?android:attr/selectableItemBackground"` so there's a ripple when it's tapped.

![Screenshot_1557999704](https://user-images.githubusercontent.com/3903757/57844250-af87e380-779d-11e9-82ea-c562b4f47d44.png)

The word "Cancel" here is confusing - it could be interpreted to mean "Cancel adding this tracking." Can we use "Don't Discard" instead?

![Screenshot_1557999691](https://user-images.githubusercontent.com/3903757/57844171-87988000-779d-11e9-9b7c-a6420a8804ec.png)

AmandaRiu commented:
> The word "Cancel" here is confusing - it could be interpreted to mean "Cancel adding this tracking." Can we use "Don't Discard" instead?

I almost feel like it should just say "YES" or "NO". @kyleaparker  WDYT?

kyleaparker commented:
>The word "Cancel" here is confusing - it could be interpreted to mean "Cancel adding this tracking." Can we use "Don't Discard" instead?

kyleaparker commented:
>The word "Cancel" here is confusing - it could be interpreted to mean "Cancel adding this tracking." Can we use "Don't Discard" instead?

This is what I would suggest: 

Are you sure you want to discard this tracking? 
- Discard
- Keep editing

We should also change the actions when exiting out of an order note to match.

anitaa1990 commented:
Changes in the PR:
* Modified the tracking events based on the P2 post. Committed in 979c916.
* Modified the discard dialog button text in this commit: c293930 and the dialog text in this commit: 0ae6ff2 (Forgot about this before 🤦‍♀ )
* Added `background="?android:attr/selectableItemBackground"` to the carrier field in 66dbc7d

* There was a bug when shipment tracking list in Order Fulfillment page where the trackings already loaded were empty the first time round but were loaded when checking again the second time. This was because the `Fulfill Order` button is clicked before the trackings were loaded in the Order detail page. And in the Order Fulfillment page, the data was being fetched from the local db.  Fixed this in 9ec5410

@AmandaRiu @nbradbury PR ready for another round of review

nbradbury commented:
@anitaa1990 This is looking great! Just two minor things I found, and then we can merge:

* Tapping a provider in the list should show a ripple.
* If you tap to select a provider, and then tap the provider that's already selected, nothing happens. I think this should just return to the previous screen

anitaa1990 commented:
@nbradbury thanks for the quick review!

@anitaa1990 This is looking great! Just two minor things I found, and then we can merge:

> Tapping a provider in the list should show a ripple.
Done in 7a2c3f1

> If you tap to select a provider, and then tap the provider that's already selected, nothing happens. I think this should just return to the previous screen.
Done in 2b44f24

nbradbury commented:
Nice work as always @anitaa1990! :shipit:
